### PR TITLE
Add an endpoint to unmerge accommodations by mergeId from the merge history

### PR DESCRIPTION
**Pull request by oleg-n-k**
Add an endpoint to unmerge accommodations by mergeId from the merge history

### CODE REVIEW

File: HappyTravel.Nakijin.Api/Controllers/AdministratorControllers/AccommodationMergeController.cs [hunk 1]
```diff
@@ -59,7 +61,28 @@ public async Task<IActionResult> MergeUncertain(int uncertainMatchId, Cancellati
         return NoContent();
     }
 
+    
+    /// <summary>
+    /// Unmerges accommodations from the Accommodation Merge History
+    /// </summary>
+    /// <param name="mergeHId">Merge ID from the Accommodation Merge History</param>
+    /// <param name="request"></param>
```
Comment (AlexeyShok) [inline HappyTravel.Nakijin.Api/Controllers/AdministratorControllers/AccommodationMergeController.cs:69]: Accommodation unmerge request

File: HappyTravel.Nakijin.Api/Controllers/AdministratorControllers/AccommodationMergeController.cs [hunk 2]
```diff
@@ -59,7 +61,28 @@ public async Task<IActionResult> MergeUncertain(int uncertainMatchId, Cancellati
...
+    /// <param name="cancellationToken"></param>
```
Comment (AlexeyShok) [inline HappyTravel.Nakijin.Api/Controllers/AdministratorControllers/AccommodationMergeController.cs:70]: Cancellation token

**Review by AlexeyShok — APPROVED** [review]

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+            if (!sourceAccommodation.IsActive)
+                return Result.Failure($"Failed to unmerge accommodations. Source accommodation '{sourceAccommodation.Id}' is deactivated");
+
+            var activeMergedAccommodations = mergedAccommodations.Where(a => a.IsActive).ToList();
+            if (activeMergedAccommodations.Any())
+                return Result.Failure(
+                    $"Merged accommodations are active already '{string.Join(", ", activeMergedAccommodations.Select(a => HtId.Create(MapperLocationTypes.Accommodation, a.Id)))}'");
+            
+            return CheckIfAccommodationSupplierCodes(sourceAccommodation, mergedAccommodations);
+
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs:123]: CheckIfContainsAccommodationSupplierCodes ?

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+            Result CheckIfAccommodationSupplierCodes(RichAccommodationDetails sourceAccommodation, List<RichAccommodationDetails> mergedAccommodations)
+            {
+                foreach (var mergedAccommodation in mergedAccommodations)
+                {
+                    var mergedAccommodationSupplierCodes = mergedAccommodation.SupplierAccommodationCodes;
+
+                    foreach (var mergedAccommodationSupplierCode in mergedAccommodationSupplierCodes)
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs:131]: maybe check by value also ?

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+            return accommodationMapping ?? Result.Failure<HtAccommodationMapping>("Failed to retrieve source accommodation mappings");
+        }
+
+
+        Result CheckIfUnmergeIsPossible(RichAccommodationDetails sourceAccommodation, List<RichAccommodationDetails> mergedAccommodations)
+        {
+            if (!sourceAccommodation.IsActive)
+                return Result.Failure($"Failed to unmerge accommodations. Source accommodation '{sourceAccommodation.Id}' is deactivated");
+
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs:116]: maybe check by Deactivation reason also ?

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+                foreach (var mergedAccommodationSupplierAndCode in mergedAccommodation.SupplierAccommodationCodes)
+                {
+                    if (sourceAccommodation.SupplierAccommodationCodes.TryGetValue(mergedAccommodationSupplierAndCode.Key, out var sourceAccommodationSupplierCode))
+                    {
+                        if (mergedAccommodationSupplierAndCode.Value == sourceAccommodationSupplierCode)
+                        {
+                            sourceAccommodation.SupplierAccommodationCodes.Remove(mergedAccommodationSupplierAndCode.Key);
+                        }
+                    }
+                }
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs:157]: needed change Deactivation reason also

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                        if (!sourceAccommodation.SupplierAccommodationCodes.TryGetValue(mergedAccommodationSupplierCode.Key, out var supplierAccommodationCode))
+                            return Result.Failure($"Failed to unmerge accommodation with {nameof(mergedAccommodation.Id)} '{mergedAccommodation.Id}. " +
+                                    $"Source accommodation doesn't contain merged supplier code '{mergedAccommodationSupplierCode.Key}");
+                    }
+                }
+
+                return Result.Success();
+            }
+        }
+
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs:142]: maybe separate into two methods and rename for understandability.

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+                .Except(mergeHistoryEvents.Select(e => e.MergedAccommodationId))
+                .ToList();
+
+            return notAssociatedIds.Any()
+                ? Result.Failure(
+                    $"{nameof(request.MergedAccommodationHtIds)} '{string.Join(", ", notAssociatedIds.Select(id => HtId.Create(MapperLocationTypes.Accommodation, id)))}' " +
+                    $"are not associated with {nameof(mergeId)} '{mergeId}'")
+                : Result.Success();
+        }
+
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs:102]: I think better to use full name of variable

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+
+        async Task<Result> ModifyAccommodationMapping(RichAccommodationDetails sourceAccommodation, List<RichAccommodationDetails> mergedAccommodations,
+            DateTimeOffset modifiedDate, CancellationToken ct)
+        {
+            var (_, getMappingFailure, accommodationMapping, getMappingError) = await GetAccommodationMapping(sourceAccommodation.Id, ct);
+            if (getMappingFailure)
+                return Result.Failure(getMappingError);
+
+            _accommodationMappingCache.Remove(accommodationMapping.MappedHtIds);
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs:189]: Here needed more actions, find existing mappings of merged accmmodations and remove Id-s from source accommodation and also activate  mappings of merged accommodations.

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+
+            accommodationMapping.Modified = modifiedDate;
+
+            _dbContext.Update(accommodationMapping);
+
+            return Result.Success();
+        }
+
+
+        async Task PublishChanges(RichAccommodationDetails sourceAccommodation, List<RichAccommodationDetails> mergedAccommodations)
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs:205]: I think no need to publish update, as I see DataCorrection not used here

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+            {
+                foreach (var mergedAccommodationSupplierAndCode in mergedAccommodation.SupplierAccommodationCodes)
+                {
+                    if (sourceAccommodation.SupplierAccommodationCodes.TryGetValue(mergedAccommodationSupplierAndCode.Key, out var sourceAccommodationSupplierCode))
+                    {
+                        if (mergedAccommodationSupplierAndCode.Value == sourceAccommodationSupplierCode)
+                        {
+                            sourceAccommodation.SupplierAccommodationCodes.Remove(mergedAccommodationSupplierAndCode.Key);
+                        }
+                    }
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs:156]: needed change IsCorrected property also

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs (hunk 4 above)
Comment (oleg-n-k) [reply HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs:131]: frequent case:
Accommodation1 contains {etg: code1}
Accommodation2 contains {etg: code2}
Accommodation3 contains {etg: code3}

Accommodations 2 and 3 have been merge to Accommodation 1
Accommodation 1 is containing only {etg: code3}
So, if we are checking by a value we cannot unmerge such cases

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs (hunk 5 above)
Comment (oleg-n-k) [reply HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs:116]: What do you mean? Unmerge accommodations only if DeactivationReson is DeactivationReasons.DuplicateInOneSupplier or MergeWithOther ?

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs (hunk 6 above)
Comment (oleg-n-k) [reply HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs:157]: DeactivationReason must be changed to none?

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs (hunk 7 above)
Comment (oleg-n-k) [reply HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs:142]: Some logic of the method has been moved to RemoveSupplierIfPossible method

File: HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs (hunk 8 above)
Comment (oleg-n-k) [reply HappyTravel.Nakijin.Api/Services/AdministratorServices/Management/Accommodations/AccommodationUnmergeService.cs:102]: mergeId is a full name.
Or what the variable do you mean?

**Pull request merged**